staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -133,20 +134,36 @@ func (s sortableWatchCacheEvents) Swap(i, j int) {
 // returned by Next() need to be events from a List() done on the underlying store of
 // the watch cache.
 // The items returned in the interval will be sorted by Key.
-func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc) (*watchCacheInterval, error) {
+func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc, key string, predicate *storage.SelectionPredicate) (*watchCacheInterval, error) {
 	buffer := &watchCacheIntervalBuffer{}
-	allItems := store.List()
+	var allItems []interface{}
+	item, exists, err := store.GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		allItems = append(allItems, item)
+	} else {
+		allItems = store.List()
+	}
 	buffer.buffer = make([]*watchCacheEvent, len(allItems))
-	for i, item := range allItems {
+	for _, item := range allItems {
 		elem, ok := item.(*storeElement)
 		if !ok {
 			return nil, fmt.Errorf("not a storeElement: %v", elem)
 		}
+		match, err := predicate.Matches(elem.Object)
```

Comment:
This is at least questionable.
The problem that I see with it is that this function is called under lock and computing the selector may potentially be expensive. There is a high chance that a customer using some non-trivial selectors will be able to hugely increase lock contention and in turn starve watch.

So even if the e2e latency is lower, the lock contention may actually be increased with this PR.

For that reason, I'm against this change.

Reply:
Make sense. I modified this PR to smaller change that only applies the key, which is a visible improvement for single object watch and should have little negative impact.